File: rust/src/bin/p0007.rs
// Project Euler: Problem 7

euler::run_solver!(7);

fn solve() -> String {
    compute(10_001).to_string()
}

fn compute(n_th: usize) -> i64 {
    use euler::math::primes;
    use std::iter;

    debug_assert!(n_th > 0);

    iter::successors(Some(2_i64), |&p| Some(primes::next_prime(p)))
        .nth(n_th - 1)
        .unwrap()
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0007_the_6th_prime() {
        assert_eq!(compute(6), 13);
    }

    #[test]
    fn p0007_the_10001th_prime() {
        assert_eq!(compute(10_001), 104743);
    }
}

File: ocaml/src/p0007.ml
(* Project Euler: Problem 7 *)

open Core

let compute n_th =
  assert (n_th > 0);

  Sequence.(
    unfold ~init:2 ~f:(fun n -> Some (n, Euler.Math.Prime.next_prime n))
    |> Fun.flip nth_exn (pred n_th))
;;

let solve () = compute 10_001 |> Int.to_string

(* Test *)

let%test_unit "6" = [%test_eq: int] (compute 6) 13
let%test_unit "10_001" = [%test_eq: int] (compute 10_001) 104743
